samza-core/src/main/scala/org/apache/samza/util/FileUtil.scala

```diff
@@ -35,13 +36,18 @@ object FileUtil {
     * */
   def writeWithChecksum(file: File, data: String): Unit = {
     val checksum = getChecksum(data)
+    val tmpFilePath = file.getAbsolutePath + ".tmp"
+    val tmpFile = new File(tmpFilePath)
     var oos: ObjectOutputStream = null
     var fos: FileOutputStream = null
     try {
-      fos = new FileOutputStream(file)
+      fos = new FileOutputStream(tmpFile)
       oos = new ObjectOutputStream(fos)
       oos.writeLong(checksum)
       oos.writeUTF(data)
+
+      //atomic swap of tmp and real offset file
+      Files.move(tmpFile.toPath, file.toPath, StandardCopyOption.ATOMIC_MOVE)
```

Comment:
1. I think we need to add `StandardCopyOption.REPLACE_EXISTING` to the existing options to `Files.move` method. `Files.move` throws up `FileAlreadyExistsException` if target file already exists and we don't provide that Replace_existing option as an additional option. 

2. Atomic move is a file system level primtiive operation and not all file system support it by default. File.move throws up `AtomicMoveNotSupportedException` if underlying file system doesn't support it.  It would be better to handle it as well. 

Please change if it makes sense(Above details are taken from Files.move javadoc).

Reply:
I don't think we need both REPLACE_EXISTING and ATOMIC_MOVE, but we might need to catch IOException as well where we catch AtomicMoveNotSupportedException. From the javadoc for ATOMIC_MOVE:

```If the target file exists then it is implementation specific if the existing file is replaced or this method fails by throwing an IOException.```